src/thor/trace_route_action.cc

```diff
@@ -242,26 +246,30 @@ void thor_worker_t::build_trace(
     Api& request) {
 
   // loop over all the segments to figure out which edge index belongs to with match result,
-  // to set the discontinuities and to remember the first and last matches we saw
+  // to set the discontinuities and to remember the first and last matches/segments we saw
   std::unordered_map<size_t, std::pair<RouteDiscontinuity, RouteDiscontinuity>> route_discontinuities;
   baldr::GraphId last_id;
   size_t edge_index = 0;
   const meili::MatchResult* origin_match = nullptr;
+  const meili::EdgeSegment* origin_segment = nullptr;
   const meili::MatchResult* dest_match = nullptr;
+  const meili::EdgeSegment* dest_segment = nullptr;
   for (const auto& path : paths) {
     // remember the global edge index
-    for (const auto& segment : path.second) {
+    for (const auto* segment : path.second) {
       if (segment->first_match_idx >= 0) {
+        if (!origin_segment) {
+          origin_segment = segment;
+        }
         match_results[segment->first_match_idx].edge_index = edge_index;
         if (!origin_match) {
           origin_match = &match_results[segment->first_match_idx];
         }
       }
       if (segment->last_match_idx >= 0) {
+        dest_segment = segment;
         match_results[segment->last_match_idx].edge_index = edge_index;
-        if (!dest_match) {
```

Comment:
this was not right. in the case of a discontinuity, we would use the wrong match result as destination